homeassistant/components/growatt_server/sensor.py

```diff
@@ -0,0 +1,123 @@
+"""Read status of growatt inverters."""
+import datetime
+import logging
+import growattServer
+import voluptuous as vol
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_NAME
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_USERNAME = "username"
+CONF_PASSWORD = "password"
+CONF_PLANT_ID = "plant_id"
+DEFAULT_UNIT = "kW"
+DEFAULT_NAME = "Growatt"
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PLANT_ID, default="0"): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Growatt sensor."""
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    plant_id = config.get(CONF_PLANT_ID)
+    name = config.get(CONF_NAME)
+
+    api = growattServer.GrowattApi()
+
+    # Log in to api and fetch first plant if no plant id is defined.
+    user_id = api.login(username, password)["userId"]
+    if plant_id == "0":
+        plant_info = api.plant_list(user_id)
+        plant_id = plant_info["data"][0]["plantId"]
+        pass
+
+    # Get a list of inverters for specified plant to add sensors for.
+    inverters = api.inverter_list(plant_id)
+
+    # Add a sensor for the total of the whole plant.
+    add_devices(
+        [GrowattInverter(api, name + "_Total", plant_id, username, password)], True
+    )
+
+    # Add sensors for each inverter in the specified plant.
+    for inverter in inverters:
+        add_devices(
+            [
+                GrowattInverter(
+                    api,
+                    name + "_" + inverter["deviceAilas"],
+                    inverter["deviceSn"],
+                    username,
+                    password,
+                )
+            ],
+            True,
+        )
+
+
+class GrowattInverter(Entity):
+    """Representation of a Growattt Sensor."""
+
+    def __init__(self, api, name, inverter_id, u, p):
```

Comment:
Please don't use single character variable names. Write out `username` and `password`.